lib/MadNLPGPU/src/interface.jl

```diff
@@ -1,12 +1,86 @@
 function MadNLP.coo_to_csc(coo::MadNLP.SparseMatrixCOO{T,I,VT,VI}) where {T,I, VT <: CuArray, VI <: CuArray}
-    csc, map = MadNLP.coo_to_csc(
-        MadNLP.SparseMatrixCOO(
-            coo.m, coo.n,
-            Array(coo.I), Array(coo.J), Array(coo.V)
-        )
+    coord = map(
+        (i,j,k)->((i,j),k),
+        coo.I, coo.J, 1:length(coo.I)
     )
+    if length(coord) > 0 
+        sort!(coord, lt = (((i, j), k), ((n, m), l)) -> (j,i) < (m,n))
+    end
+    
+    mapptr = getptr(CUDABackend(), coord)
+
+    colptr = similar(coo.I, size(coo,2)+1)
+    
+    coord_csc = coord[@view(mapptr[1:end-1])]
+
+    ker_colptr!(CUDABackend())(colptr, coord_csc, ndrange = length(coord_csc))
+    rowval = map(x -> x[1][1], coord_csc)
+    nzval = similar(rowval, T)
+
+    csc = CUDA.CUSPARSE.CuSparseMatrixCSC(colptr, rowval, nzval, size(coo))
+    
+    
+    cscmap = similar(coo.I, Int)
+    ker_index(CUDABackend())(cscmap, mapptr, coord, ndrange = length(mapptr)-1)
+    
+    synchronize(CUDABackend())
+    
+    return csc, cscmap
+
+end
 
-    return CUDA.CUSPARSE.CuSparseMatrixCSC(csc), CuArray(map)
+@kernel function ker_colptr!(colptr, @Const(coord))
+    index = @index(Global)
+
+    @inbounds begin
+        if index == 1
+            ((i2,j2),k2) = coord[index]
+            colptr[1:j2] .= 1
+        else
+            ((i1,j1),k1) = coord[index-1]
+            ((i2,j2),k2) = coord[index]
+            if j1 != j2
+                colptr[j1+1:j2] .= index 
+            end
+            if index == length(coord)
+                colptr[j2+1:end] .= index+1
+            end
+        end
+    end
+end
+
+@kernel function ker_index(cscmap, @Const(mapptr),@Const(coord))
+    index = @index(Global)
+    @inbounds for l in mapptr[index]:mapptr[index+1]-1
+        ((i,j),k) = coord[l]
+        cscmap[k] = index
+    end        
+end
+
+function getptr(backend, array; cmp = isequal)
+
+    bitarray = similar(array, Bool, length(array) + 1)
+    kergetptr(backend)(cmp, bitarray, array; ndrange = length(array) + 1)
+    synchronize(backend)
+
+    return findall(identity, bitarray)
+end
+@kernel function kergetptr(@Const(cmp), bitarray, @Const(array))
```

Comment:
It turned out that it is redundant to 
https://github.com/MadNLP/MadNLP.jl/blob/master/src/KKT/sparse.jl#L597-L602

Just have replaced it with the existing one. 

This function compares two consecutive data entry of an array based on the given criteria, and returns the indices where they do not match. This can be used, e.g., for computing `colptr` of `SparseMatrixCSC`